app/src/org/commcare/android/tasks/ManageKeyRecordTask.java

```diff
@@ -37,15 +39,17 @@
  * This task uses three steps
  * 1) Clean up user key records and figure out whether we need to look for
  * new records
- * 2) Fetch new records [HTTP step]
+ * 2) Fetch new records [HTTP step] (not always executed)
  * 3) Process the new records and perform any necessary data migration
  * 
  * @author ctsims
  *
  */
 public abstract class ManageKeyRecordTask<R> extends HttpCalloutTask<R> {
-    final String username;
-    final String password;
+    private final String username;
+    private String password;
+    private String pin;
+    private final LoginActivity.LoginMode loginMode;

```

Comment:
I would probably introduce this coupling in the other direction or put it somewhere neutral (Just moving the enum definition that is). In general it's better it Tasks don't 'know' about activities